sqlglot/optimizer/scope.py

```diff
@@ -0,0 +1,274 @@
+
+
+def _traverse_scope(scope):
+    # Collect all the relevant Expression parts in this scope
+    for node, parent, _ in _bfs_until_next_scope(scope.expression):
+        if node is scope.expression:
+            continue  # Skip this node itself - we only care about children
+        if isinstance(node, (exp.Select, exp.Union)):
+            if isinstance(parent, exp.CTE):
+                scope.ctes.append(parent)
+            elif isinstance(parent, exp.Subquery):
+                scope.derived_tables.append(parent)
+            else:
+                scope.subqueries.append(node)
+        elif isinstance(node, exp.Table):
+            scope.tables.append(node)
+        elif isinstance(node, exp.Column) and not isinstance(node.this, exp.Star):
+            scope.columns.append(node)
+
+    if isinstance(scope.expression, exp.Select):
+        yield from _traverse_select(scope)
+    elif isinstance(scope.expression, exp.Union):
+        yield from _traverse_union(scope)
+    else:
+        raise OptimizeError(f"Unexpected expression type: {scope.expression}")
+    yield scope
+
+
+def _traverse_select(scope):
+    yield from _traverse_derived_tables(scope.ctes, scope)
+    yield from _traverse_derived_tables(scope.derived_tables, scope)
+
+    for table in scope.tables:
+        table_name = table.text("this")
+
+        if isinstance(table.parent, exp.Alias):
+            selectable_table = table.parent.alias
+        else:
+            selectable_table = table_name
+
+        if table_name in scope.selectables:
+            # This is a reference to a parent selectable (e.g. a CTE), not an actual table.
+            scope.selectables[selectable_table] = scope.selectables[table_name]
+        elif selectable_table in scope.selectables:
+            raise OptimizeError(f"Duplicate table name: {selectable_table}")
+        else:
+            scope.selectables[selectable_table] = table
+
+    yield from _traverse_subqueries(scope)
+
+
+def _traverse_union(scope):
+    if scope.ctes:
+        yield from _traverse_derived_tables(scope.ctes, scope)
+
+    # The last scope to be yield should be the top most scope
+    top_left = None
+    for child_scope in _traverse_scope(scope.branch(scope.expression.left)):
+        yield child_scope
+        top_left = child_scope
+
+    top_right = None
+    for child_scope in _traverse_scope(scope.branch(scope.expression.right)):
+        yield child_scope
+        top_right = child_scope
+
+    scope.union = (top_left, top_right)
+
+
+def _traverse_derived_tables(derived_tables, scope):
+    for derived_table in derived_tables:
+        for child_scope in _traverse_scope(
+            scope.branch(
+                derived_table.this, outer_column_list=derived_table.alias_column_names
+            )
+        ):
+            yield child_scope
+            # Tables without aliases will be set as ""
+            scope.selectables[derived_table.alias] = child_scope
```

Comment:
ok to have duplicates here clashing?

Reply:
Hm, I'm not sure what standard SQL behavior is.

Seems like sqlite allows duplicate selectable names... it just throws an error if you reference an ambiguous column.

That means we'd have to track all selectables, including duplicates. Hm